src/Components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,131 @@
+import React, { useContext, useEffect, useRef, useState } from 'react';
+import classNames from 'classnames';
+import { Todo } from '../../types/Todo';
+import { TodosContext } from '../../TodosContext';
+
+type Props = {
+  todo: Todo;
+};
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const { todos, setTodos, deleteTodo, toggleTodoCompleted, editTodo } =
+    useContext(TodosContext);
+  const [editing, setEditing] = useState(false);
+  const [editingValue, setEditingValue] = useState(todo.title);
+  const [editingId, setEditingId] = useState('');
+
+  const titleField = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (titleField.current) {
+      titleField.current.focus();
+    }
+  }, [editingId, editing]);
+
+  const handleTodoCompleted = (id: number, completed: boolean) => {
+    toggleTodoCompleted(id, completed);
+    const updateTodos = todos.map(currentTodo => {
+      if (currentTodo.id === id) {
+        return { ...currentTodo, completed: !currentTodo.completed };
+      }
+
+      return currentTodo;
+    });
+
+    setTodos(updateTodos);
+  };
+
+  const todoEdit = (id: number, title: string) => {
+    const currentTitle = title.trim();
+
+    editTodo(id, currentTitle);
+    setEditing(false);
+    setEditingId(todo.id.toString());
+
+    const editingTodos = todos.map(item => {
+      if (todo.id === item.id) {
+        return { ...item, title: currentTitle };
+      }
+
+      return item;
+    });
+
+    setTodos(editingTodos);
+  };
+
+  return (
+    <div
+      key={todo.id.toString()}
+      data-cy="Todo"
+      className={classNames('todo', {
+        editing: editing,
+        completed: todo.completed,
+      })}
+      onDoubleClick={() => {
+        setEditing(true);
+        setEditingId(todo.id.toString());
+      }}
```

Comment:
Pls move the function inside `onDoubleClick` handler to a separate const and use as a callback here
**PLS DO THE SAME FOR ALL SUCH CASES IN THE PROJECT**